enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+---
+title: upgrade-without-registry
+authors:
+- "@jhernand"
+reviewers:
+- "@avishayt"
+- "@danielerez"
+- "@mrunalp"
+- "@nmagnezi"
+- "@oourfali"
+approvers:
+- "@sdodson"
+- "@zaneb"
+api-approvers:
+- "@sdodson"
+- "@zaneb"
+creation-date: 2023-06-29
+last-updated: 2023-07-26
+tracking-link:
+- https://issues.redhat.com/browse/RFE-4482
+see-also:
+- https://issues.redhat.com/browse/OCPBUGS-13219
+- https://github.com/openshift/cluster-network-operator/pull/1803
+replaces: []
+superseded-by: []
+---
+
+# Upgrade without registry
+
+## Summary
+
+Provide an automated mechanism to upgrade a cluster without requiring an image
+registry server.
+
+## Motivation
+
+All these stories are in the context of disconnected clusters with limited
+resources, both in the cluster itself and in the surrounding environment:
+
+- The cluster is not connected to the Internet, or the bandwidth is very limited.
+- It isn't possible to bring up additional machines, even temporarily.
+- The resources of the cluster are limited, in particular in terms of CPU and memory.
+- The technicians responsible for performing the upgrade have no OpenShift knowledge.
+
+These clusters are usually installed at the customer site by a partner engineer
+collaborating with customer technicians.
+
+Eventually, the cluster will need to be upgraded, and then the technicians will
+need tools that make the process as simple as possible, ideally requiring no
+OpenShift knowledge.
+
+### User Stories
+
+#### Prepare an upgrade bundle
+
+As an engineer working in a partner's factory, I want to be able to assemble an
+upgrade bundle that contains all the artifacts (container images and metadata)
+needed to upgrade the OpenShift cluster and my own applications. I want to hand
+over this upgrade bundle and the documentation explaining how to use it to the
+technicians that will perform the upgrade in a suitable media, for example a USB
+stick.
+
+#### Include custom images in the bundle
+
+As an engineer working in a partner's factory I want to be able to include in
+the upgrade bundle custom images for workloads specific to the customer.
+
+#### Explicitly allow vetted upgrade bundle
```

Comment:
Is there any requirement to handle signatures of bundles to ensure they have not been tampered with?

Reply:
Partners will definitely ask for this.

Reply:
There is already an optional `digest` field to ensure the integrity of the bundle. This could be a signature, if required. The idea is that the engineer that administrates the cluster would set this field to the digest of the bundle that she is explicitly approving. Then, if the technician makes a mistake (uses the wrong bundle) or if the bundle has been tampered with the cluster version operator will reject it and abort the upgrade.

Note also that even without this digest (or signature) of the bundle there is still the signature of the release image itself: if the release payload has been tampered with the upgrade will be rejected. The additional digest (or signature) is intended to detect the problem earlier, as well as to protect the custom images that aren't part of the OpenShift release.

We could use a signature instead of a digest, but that would complicate things for partners because they are the ones that build the bundle, and the ones that would have to sign it. Using a signature would force them to also distribute the corresponding keys to verify that signature.

Reply:
We count on CVO verifying the signature, but is not enough because we also want to include customer images and metadata in the bundle: the digest covers all that.

CVO doesn't have direct access to the bundle, only the `bundle-extractor` pods running in each node have that access. These pods will check the digest on the fly, as the bundle is extracted. Once finished they report back via conditions in the `ClusterVersion` object. CVO checks these conditions.

Reply:
Yes, that is the idea of the digest: it covers the complete bundle, including the metadata.

I will add a paragraph explaining that the bundle should also contain the payload signatures.